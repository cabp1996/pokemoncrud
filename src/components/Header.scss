.search-bar-container {
  display: flex;
  justify-content: space-between;

  .search-input {
    position: relative;

    svg {
      position: absolute;
      bottom: -1;
      left: 50;
    }

    input {
      box-sizing: border-box;
      padding-left: 30px !important;
    }
  }
}
