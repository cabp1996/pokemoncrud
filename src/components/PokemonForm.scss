.form-container {
  border: 2px solid lightgrey;
  border-radius: 4px;
  box-sizing: border-box;
  margin-top: 1.5rem;
  padding: 1rem 4rem;

  h4 {
    margin: 0;
    margin-bottom: 1.5rem;
  }

  .title {
    text-align: center;
  }

  .form-grid {
    display: grid;
    grid-template-columns: repeat(2, 1fr);
    column-gap: 48px;
    row-gap: 48px;
  }

  .form-control {
    display: grid;
    grid-template-columns: 30% 70%;
    column-gap: 24px;
    align-items: center;

    .range-input {
      //range input
    }
  }

  .buttons-container {
    margin-top: 1.5rem;
    display: flex;
    justify-content: center;
  }
}
