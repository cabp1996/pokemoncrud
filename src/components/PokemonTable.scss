.pokemon-table {
  width: 100%;
  border-collapse: collapse;
  margin-top: 1.5rem;

  td,
  th {
    border: 1px solid grey;
    text-align: center;
  }

  th {
    box-sizing: border-box;
    padding: 0.5rem 1rem;
  }

  img {
    height: 50px;
  }

  .btn-icon {
    background-color: white;
    border: none;
    color: var(--main-color);
    cursor: pointer;
    font-size: 16px;
    font-weight: bold;
    outline: none;
  }

  .td-actions {
    display: grid;
    justify-content: space-evenly;
    grid-template-columns: repeat(2, 1fr);
    grid-auto-flow: column;
    align-items: center;
    justify-items: center;
  }
}
