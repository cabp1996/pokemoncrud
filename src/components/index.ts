export * from "./Header";
export * from "./PokemonForm";
export * from "./PokemonTable";
