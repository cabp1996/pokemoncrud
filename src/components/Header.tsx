import { useDispatch } from "react-redux";
import { showPokemonForm, searchPokemon } from "../actions";
import { useEffect, useState } from "react";
import "./Header.scss";

export const Header = () => {
  const dispatch = useDispatch();

  const [filter, setFilter] = useState<string>("");

  useEffect(() => {
    dispatch(searchPokemon(filter));
  }, [filter]);

  const handleChangeSearch = (e: React.ChangeEvent<HTMLInputElement>) => {
    setFilter(e.target.value);
  };

  const handleShowPokemonForm = () => dispatch(showPokemonForm());

  return (
    <>
      <h1>Listado de Pokemon</h1>
      <div className="search-bar-container">
        <div className="search-input">
          <svg
            viewBox="0 0 24 24"
            width="24"
            height="24"
            stroke="grey"
            strokeWidth="1"
            fill="none"
          >
            <circle cx="11" cy="11" r="8"></circle>
            <line x1="21" y1="21" x2="16.65" y2="16.65"></line>
          </svg>
          <input
            className="form-input"
            placeholder="Buscar"
            type="text"
            value={filter}
            name="filter"
            onChange={handleChangeSearch}
          />
        </div>

        <button className="main-button" onClick={handleShowPokemonForm}>
          <i className="fa fa-plus" aria-hidden="true"></i> Nuevo
        </button>
      </div>
    </>
  );
};
