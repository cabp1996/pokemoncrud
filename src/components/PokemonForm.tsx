import { useDispatch, useSelector } from "react-redux";

import { RootState } from "../reducers/root.reducer";
import { useForm } from "../hooks";
import { useEffect } from "react";
import {
  hidePokemonForm,
  requestCreatePokemonAction,
  requestUpdatePokemonAction,
} from "../actions";

import { Pokemon } from "../interfaces";
import "./PokemonForm.scss";

export const PokemonForm = () => {
  const dispatch = useDispatch();

  const { selectedPokemon } = useSelector((state: RootState) => state.pokemons);

  const [formValues, handleInputChange, setValues] = useForm({
    name: "",
    image: "",
    attack: 0,
    defense: 0,
  });

  const { name, image, attack, defense } = formValues;

  useEffect(() => {
    if (selectedPokemon) {
      setValues({
        name: selectedPokemon.name,
        image: selectedPokemon.image,
        attack: selectedPokemon.attack,
        defense: selectedPokemon.defense,
      });
    } else {
      setValues({
        name: "",
        image: "",
        attack: 0,
        defense: 0,
      });
    }
  }, [selectedPokemon, setValues]);

  const handleHideForm = () => {
    dispatch(hidePokemonForm());
  };

  const handleSubmitForm = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    const pokemonToSave: Pokemon = {
      name,
      attack: Number(attack),
      defense: Number(defense),
      hp: 100,
      idAuthor: 1,
      image,
      type: "normal",
    };

    if (selectedPokemon === null) {
      dispatch(requestCreatePokemonAction(pokemonToSave));
    } else {
      dispatch(requestUpdatePokemonAction(selectedPokemon.id, pokemonToSave));
    }
  };

  const isSaveButtonDisabled = (): boolean => {
    return name === "" || image === "";
  };

  return (
    <div className="form-container">
      <div className="title">
        <h4>{selectedPokemon ? "Editar pokemon" : "Nuevo pokemon"}</h4>
      </div>

      <form onSubmit={handleSubmitForm}>
        <div className="form-grid">
          <div className="form-control">
            <label htmlFor="inputName">Nombre:</label>
            <input
              className="form-input"
              id="inputName"
              type="text"
              name="name"
              value={name}
              onChange={handleInputChange}
            />
          </div>

          <div className="form-control">
            <label htmlFor="inputAttack">Ataque:</label>
            <input
              id="inputAttack"
              className="range-input"
              type="range"
              min="0"
              max="100"
              name="attack"
              value={attack}
              onChange={handleInputChange}
            />
          </div>

          <div className="form-control">
            <label htmlFor="inputImage">Imagen:</label>
            <input
              className="form-input"
              id="inputImage"
              type="text"
              name="image"
              value={image}
              onChange={handleInputChange}
            />
          </div>

          <div className="form-control">
            <label htmlFor="inputDefense">Defensa:</label>
            <input
              id="inputDefense"
              className="range-input"
              type="range"
              min="0"
              max="100"
              name="defense"
              value={defense}
              onChange={handleInputChange}
            />
          </div>
        </div>

        <div className="buttons-container">
          <button
            className="main-button mr-8"
            disabled={isSaveButtonDisabled()}
            type="submit"
          >
            <i className="fa fa-floppy-o" aria-hidden="true"></i> Guardar
          </button>
          <button className="main-button" onClick={handleHideForm}>
            <i className="fa fa-times" aria-hidden="true"></i> Cancelar
          </button>
        </div>
      </form>
    </div>
  );
};
