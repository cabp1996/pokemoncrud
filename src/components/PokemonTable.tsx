import { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import {
  requestDeletePokemonAction,
  requestGetPokemonsListAction,
  selectPokemonAction,
} from "../actions";
import { RootState } from "../reducers/root.reducer";
import { Pokemon } from "../interfaces";
import "./PokemonTable.scss";

export const PokemonTable = () => {
  const dispatch = useDispatch();

  const [filteredPokemons, setFilteredPokemons] = useState<Pokemon[]>([]);

  const { pokemons, filterSearch } = useSelector(
    (state: RootState) => state.pokemons
  );

  useEffect(() => {
    dispatch(requestGetPokemonsListAction());
  }, [dispatch]);

  useEffect(() => {
    setFilteredPokemons(pokemons);
  }, [pokemons]);

  useEffect(() => {
    if (pokemons.length > 0) {
      const filtered: Pokemon[] = pokemons.filter((pokemon: Pokemon) =>
        pokemon.name.toLowerCase().includes(filterSearch.toLocaleLowerCase())
      );
      setFilteredPokemons(filtered);
    }
  }, [filterSearch, pokemons]);

  const handleSelectPokemonToUpdate = (pokemon: Pokemon) => {
    dispatch(selectPokemonAction(pokemon));
  };

  const handleDeletePokemon = async (id: number) => {
    dispatch(requestDeletePokemonAction(id));
  };

  return (
    <table className="pokemon-table">
      <thead>
        <tr>
          <th>Nombre</th>
          <th>Imagen</th>
          <th>Ataque</th>
          <th>Defensa</th>
          <th>Acciones</th>
        </tr>
      </thead>
      <tbody>
        {filteredPokemons.map((pokemon, index) => (
          <tr key={index}>
            <td>{pokemon.name}</td>
            <td>
              <img
                src={pokemon.image ? pokemon.image : "/images/notfound.png"}
                alt={pokemon.name}
              />
            </td>
            <td>{pokemon.attack}</td>
            <td>{pokemon.defense}</td>
            <td>
              <div className="td-actions">
                <button
                  className="btn-icon"
                  onClick={() => handleSelectPokemonToUpdate(pokemon)}
                >
                  <i className="fa fa-pencil" aria-hidden="true"></i>
                </button>

                <button
                  className="btn-icon"
                  onClick={() => handleDeletePokemon(pokemon.id)}
                >
                  <i className="fa fa-trash" aria-hidden="true"></i>
                </button>
              </div>
            </td>
          </tr>
        ))}
      </tbody>
    </table>
  );
};
