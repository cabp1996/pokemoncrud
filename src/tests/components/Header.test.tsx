import configureStore from "redux-mock-store";
import { initialState } from "../mocks/initial-state.mock";
import { Provider } from "react-redux";
import { Header } from "../../components";
import { fireEvent, render, RenderResult } from "@testing-library/react";
import thunk from "redux-thunk";

jest.mock("../../actions", () => ({
  showPokemonForm: jest.fn(),
  searchPokemon: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
let store = mockStore(initialState);

store.dispatch = jest.fn();

describe("Header tests", () => {
  let wrapper: RenderResult;

  beforeEach(() => {
    wrapper = render(
      <Provider store={store}>
        <Header />
      </Provider>
    );
  });

  it("should match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should new button call dispatch action to show pokemon form", async () => {
    const spy = jest.spyOn(store, "dispatch");
    const button = await wrapper.getByRole("button");
    button.click();
    expect(spy).toHaveBeenCalled();
  });

  it("should onChange search filter input dispatch filter text", async () => {
    const spy = jest.spyOn(store, "dispatch");
    const searchInput = wrapper.container.querySelector(
      ".form-input"
    ) as HTMLInputElement;

    fireEvent.change(searchInput, { target: { value: "pokemon" } });

    expect(spy).toHaveBeenCalled();
  });
});
