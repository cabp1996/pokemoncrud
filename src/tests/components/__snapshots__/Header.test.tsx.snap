// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header tests should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h1>
        Listado de Pokemon
      </h1>
      <div
        class="search-bar-container"
      >
        <div
          class="search-input"
        >
          <svg
            fill="none"
            height="24"
            stroke="grey"
            stroke-width="1"
            viewBox="0 0 24 24"
            width="24"
          >
            <circle
              cx="11"
              cy="11"
              r="8"
            />
            <line
              x1="21"
              x2="16.65"
              y1="21"
              y2="16.65"
            />
          </svg>
          <input
            class="form-input"
            name="filter"
            placeholder="Buscar"
            type="text"
            value=""
          />
        </div>
        <button
          class="main-button"
        >
          <i
            aria-hidden="true"
            class="fa fa-plus"
          />
           Nuevo
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <h1>
      Listado de Pokemon
    </h1>
    <div
      class="search-bar-container"
    >
      <div
        class="search-input"
      >
        <svg
          fill="none"
          height="24"
          stroke="grey"
          stroke-width="1"
          viewBox="0 0 24 24"
          width="24"
        >
          <circle
            cx="11"
            cy="11"
            r="8"
          />
          <line
            x1="21"
            x2="16.65"
            y1="21"
            y2="16.65"
          />
        </svg>
        <input
          class="form-input"
          name="filter"
          placeholder="Buscar"
          type="text"
          value=""
        />
      </div>
      <button
        class="main-button"
      >
        <i
          aria-hidden="true"
          class="fa fa-plus"
        />
         Nuevo
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
